text/0000-ember-data-identifiers.md

```diff
@@ -0,0 +1,396 @@
+ will be discussed further below.
+
+```typescript
+/*
+  A method which can expect to receive various data as it's first argument
+  and the name of a bucket as it's second argument. Currently the second
+  argument will always be `record` data will always conform to a `json-api`
+  `Resource` interface.
+
+  The method must return a unique (to at-least the given bucket) string identifier
+  for the given data as a string to be used as the `lid` of an `Identifier` token.
+
+  This method will only be called when an identifier for the supplied data is
+  not already known via `lid` or `type + id` combo and one needs to be generated
+  or retrieved from a proprietary cache.
+*/
+type GenerationMethod = (data: Object, bucket: string) => string;
+
+export function setIdentifierGenerationMethod(method: GenerationMethod): void;
+```
+
+A simple custom generation method might be an increasing counter like below:
+
+```typescript
+import { setIdentifierGenerationMethod } form '@ember-data/store';
+let count = 0;
+
+setIdentifierGenerationMethod((resource: Resource) => {
+  return resource.lid || `my-key-${count++}`;
+});
+```
+
+### Identifiers for Records
+
+When discussing identifiers for records it is useful to be familiar with `json-api` interfaces for [ResourceObjects](https://jsonapi.org/format/#document-resource-objects) and [ResourceIdentifierObjects](https://jsonapi.org/format/#document-resource-identifier-objects).
+
+Below, we expose a rough approximation of these interfaces as `Resource` including the potential presence of `lid`.
+
+```typescript
+import { Value as JSONValue } from 'json-typescript';
+
+type JSONDict = { [k: string]: JSONValue };
+
+export interface Resource {
+  id: string;
+  type: string;
+  lid?: string;
+  attributes?: JSONDict;
+  relationships?: JSONDict;
+  meta?: JSONDict;
+}
+```
+
+We can access and generate identifiers for records using the following APIs available on the `Store`
+ and `StoreWrapper` classes.
+
+```typescript
+export interface Store {
+  /*
+   Returns the Identifier for the given Resource, creates one if it does not yet exist.
+
+   Specifically this means that we:
+
+   - validate the `id` `type` and `lid` combo against known identifiers
+   - return an object with an `lid` that is stable (repeated calls with the same
+    `id` + `type` or `lid` will return the same `lid` value)
+   - this referential stability of the object itself is guaranteed
+  */
+  getOrCreateRecordIdentifier(resource: Resource): Identifier {}
+
+  /*
+   Returns a new Identifier for the supplied data. Call this method to generate
+   an identifier when  a new resource is being created local to the client and
+   potentially does not have an `id`.
+  */
+  createIdentifierForNewRecord(type: string): Identifier {}
+
+  updateRecordIdentifier(identifier: Identifier, data: Resource): void;
+
+  forgetRecordIdentifier(identifier: Identifier): void
```

Comment:
`forget` differs significantly from `unload` in semantics and a new verb was chosen here in part to convey intent and partly to avoid the baggage that comes with `unload`.

`forget` is intended to fully purge a given identifier from the identifier cache. It differs in semantics from `unload` primarily in how we consider the existence (or lack thereof) of remote data.

What this means is that we treat the identifier as if we no longer know anything about it locally or remotely. E.g. it may or may not exist remotely still, but we remove all reference to it.

This differs from the semantics of `unload` in which if a record has not first had a deletion persisted we move to a state in which we assume no local data but that remote data does still exist.

Reply:
ok.  I'm not a fan of the name.  I would recommend `removeRecordIdentifier` mainly b/c `remove` is more standard terminology for caches than `forget`.

I don't plan on being a 1-man bikeshedder though if others feel strongly about `forget`.